src/pull_request.py

```diff
@@ -0,0 +1,230 @@
+
+    Raises:
+        GitError if invalid remote url.
+
+    Returns:
+        Git repository name. e.g. canonical/upload-charm-docs.
+    """
+    matched_repository = HTTPS_URL_PATTERN.match(remote_url)
+    if not matched_repository:
+        raise GitError(f"Invalid remote repository name {remote_url=!r}")
+    return matched_repository.group(1)
+
+
+def create_github(access_token: typing.Any):
+    """Create a Github instance to handle communication with Github server.
+
+    Args:
+        access_token: Access token that has permissions to open a pull request.
+
+    Raises:
+        InputError: if invalid token format input.
+
+    Returns:
+        A Github repository instance.
+    """
+    if not access_token:
+        raise InputError(
+            f"Invalid 'access_token' input, it must be non-empty, got {access_token=!r}"
+        )
+    if not isinstance(access_token, str):
+        raise InputError(
+            f"Invalid 'access_token' input, it must be a string, got {access_token=!r}"
+        )
+
+    return Github(login_or_token=access_token)
+
+
+def create_pull_request(
+    repository: Repo,
+    github_repository: Repository,
+    branch_name: str | None,
+    dry_run: bool,
+) -> str:
+    """Create pull request for changes in given repository path.
+
+    Args:
+        access_token: Github access token.
+        repository_path: Repository root where .git resides.
+        branch_name: Pull request branch name.
+        dry_run: If enabled, only log the action that would be taken.
+
+    Raises:
+        InputError: if branch name configuration is invalid.
+
+    Returns:
+        Pull request URL string. None if no pull request was created/modified.
+    """
+    branch_name = branch_name or DEFAULT_BRANCH_NAME
+    base = repository.active_branch.name
+    if base == branch_name:
+        raise InputError("Branch name cannot be equal to base branch.")
+
+    if not repository.is_dirty(untracked_files=True):
+        return PR_LINK_NO_CHANGE
+
+    _configure_user(repository=repository)
+
+    if _check_branch_exists(repository=repository, branch_name=branch_name):
+        _merge_existing_branch(
+            repository=repository,
+            branch_name=branch_name,
+            commit_msg=ACTIONS_COMMIT_MESSAGE,
+            dry_run=dry_run,
+        )
+    else:
+        _create_branch(
+            repository=repository,
+            branch_name=branch_name,
+            commit_msg=ACTIONS_COMMIT_MESSAGE,
+            dry_run=dry_run,
+        )
+    repository.git.checkout(base)
+
+    open_pulls = github_repository.get_pulls(
+        state="open", head=f"{github_repository.full_name}:{branch_name}"
+    )
+    if not list(open_pulls):
+        pr_url = _create_pull_request(
+            github_repository=github_repository,
+            branch_name=branch_name,
+            base=base,
+            dry_run=dry_run,
+        )
+    else:
+        pr_url = open_pulls[0].html_url
```

Comment:
Would it be worth putting a comment here noting that there can only be on PR per branch? And this means that the one at index 0 is the one we generated?